### ISSUE TITLE
Update cid and multi hash

### ISSUE THREAD

**Issue by jennijuju** [filecoin-project/ref-fvm#1233]
#1211 reverted the pkg update, we should introduce them back.

however, given the original committer, forest team has confirmed this update is not too urgent - we will do this in a later release, i.e: post nv18 release

**Comment by lemmih** [comment]
@jennijuju @Stebalien @shawnrader This is becoming an increasingly urgent issue for Forest as we cannot work on NV18 before this can been resolved. Are there any blockers for completing this task by next week? If there _are_ any blockers, let's schedule a meeting to discuss a way forward.

**Comment by vmx** [comment]
It's in the works: https://github.com/filecoin-project/filecoin-ffi/pull/355

**Comment by lemmih** [comment]
> It's in the works: [filecoin-project/filecoin-ffi#355](https://github.com/filecoin-project/filecoin-ffi/pull/355)

That won't work for Forest. We need FVM2 to be updated.

**Comment by lemmih** [comment]
> however, given the original committer, forest team has confirmed this update is not too urgent - we will do this in a later release, i.e: post nv18 release

Oops, I think there's been a miscommunication here. When we talked in early December, I meant it wasn't so urgent that it had to be done before the Christmas break. It definitely has to happen before the nv18 release.

**Comment by lemmih** [comment]
To add some context for those who weren't part of the discussion(s) in December:

- The CID and multihash dependencies have to be updated in FVM2.
- It is a show-stopper for Forest if this doesn't happen.
- Updating FVM2 is trivial (Forest already submitted a patch for doing this) but, due to an unfortunately versioning scheme, it will break a lot of downstream projects (including filecoin-ffi).
- It was agreed that FVM2 would get a new semver-violating minor release with updated CID/multihash dependencies.

**Comment by shawnrader** [comment]
Filecoin-ffi as it is now should not have any issues with updating cid and multihash crates as long as FVM3 and FVM2 are using the same version of those crates.

**Comment by lemmih** [comment]
> Filecoin-ffi as it is now should not have any issues with updating cid and multihash crates as long as FVM3 and FVM2 are using the same version of those crates.

Excellent! This means we're ready to bump the CID/multihash dependencies in FVM2?

**Comment by jennijuju** [comment]
> > Filecoin-ffi as it is now should not have any issues with updating cid and multihash crates as long as FVM3 and FVM2 are using the same version of those crates.
> 
> Excellent! This means we're ready to bump the CID/multihash dependencies in FVM2?

This is being worked on here https://github.com/filecoin-project/filecoin-ffi/pull/355 - but IIRC the dependency mentioned  here is out of scope?

**Comment by jennijuju** [comment]
> > > Filecoin-ffi as it is now should not have any issues with updating cid and multihash crates as long as FVM3 and FVM2 are using the same version of those crates.
> > 
> > 
> > Excellent! This means we're ready to bump the CID/multihash dependencies in FVM2?
> 
> This is being worked on here [filecoin-project/filecoin-ffi#355](https://github.com/filecoin-project/filecoin-ffi/pull/355) - but IIRC the dependency mentioned here is out of scope for nv18 and forest was okay with that? Source https://filecoinproject.slack.com/archives/C029MT4PQB1/p1670481960248989?thread_ts=1670435116.931999&channel=C029MT4PQB1&message_ts=1670481960.248989

**Comment by lemmih** [comment]
@jennijuju I think we must have had a different interpretation of "nv18 code freeze". I thought that meant "nv18 feature freeze" (ie January) rather than "nv18 release" (ie March or later). When I discussed the matter with Shawn Rader, I told him we didn't need to resolve it in December and that resolving it in January would be fine.

On a side note, I don't understand the need for https://github.com/filecoin-project/filecoin-ffi/pull/355. In the future, FVM2 and FVM3 will always use the same versions of `cid` and `multihash`.

**Comment by lemmih** [comment]
Miscommunications aside, what matters is that this is a blocker for Forest. We both agree that this work needs to be done quite soon (I would like it done this month, you thought it was gonna be done in March). Let's bite the bullet and just do the work now rather than later. The Forest team is more than willing to provide engineers to handle the Rust side of things. You'll then have to update the Lotus-related projects that are downstream of the FVM.

Let's have a call early next week to make sure we're on the same page.

**Comment by jennijuju** [comment]
Update: the current cid/multihash version  is aligned between fvm v2.2.0 and fvm v3 and that unblocks forest for nv18.
more details see [here](https://filecoinproject.slack.com/archives/C029MT4PQB1/p1673706826673439?thread_ts=1673705328.883269&channel=C029MT4PQB1&message_ts=1673706826.673439) 

Forest needs cid 0.9 for a QoL improvement, thus would be nice to get this done post nv18 soon.

**Comment by lemmih** [comment]
Can Forest assist with this upgrade? It's becoming a pressing issue for us. Because of this, we cannot upgrade several other dependencies (e.g., libipld-core`). If upgrading FVM2 and FVM3 isn't possible in the next quarter or two, Forest will have to fork the FVM.

**Comment by Stebalien** [comment]
This has been fixed and released for both fvm v2 and fvm v3. Furhtermore, both fvm v2/v3 now use the same `fvm_ipld_encoding`, `fvm_ipld_blockstore`, hamt, amt, etc. libraries.

Versions:

- v2
    - fvm v2.5.0
    - fvm_shared v2.5.0
- v3
    - fvm v3.5.0
    - fvm_shared v3.4.0
    - fvm_integration_tests v3.1.0
- other:
    - fvm_ipld_kamt v0.3.0
    - fvm_ipld_hamt v0.7.0
    - fvm_ipld_amt v0.6.0
    - fvm_ipld_encoding v0.4.0
    - fvm_ipld_blockstore v0.2.0
    - fvm_ipld_car v0.7.0
